pkg/crc/machine/vsock.go

```diff
@@ -72,51 +74,51 @@ func listOpenPorts(daemonClient *daemonclient.Client) ([]types.ExposeRequest, er
 
 const (
 	virtualMachineIP = "192.168.127.2"
-	internalSSHPort  = 22
+	internalSSHPort  = "22"
 	localIP          = "127.0.0.1"
-	httpPort         = 80
-	httpsPort        = 443
-	apiPort          = 6443
-	cockpitPort      = 9090
+	httpPort         = "80"
+	httpsPort        = "443"
+	apiPort          = "6443"
+	cockpitPort      = "9090"
 )
 
 func vsockPorts(preset crcPreset.Preset) []types.ExposeRequest {
 	exposeRequest := []types.ExposeRequest{
 		{
 			Protocol: "tcp",
-			Local:    fmt.Sprintf("%s:%d", localIP, constants.VsockSSHPort),
-			Remote:   fmt.Sprintf("%s:%d", virtualMachineIP, internalSSHPort),
+			Local:    net.JoinHostPort(localIP, strconv.Itoa(constants.VsockSSHPort)),
+			Remote:   net.JoinHostPort(virtualMachineIP, internalSSHPort),
 		},
 	}
 	switch preset {
 	case crcPreset.OpenShift:
 		exposeRequest = append(exposeRequest,
 			types.ExposeRequest{
 				Protocol: "tcp",
-				Local:    fmt.Sprintf("%s:%d", localIP, apiPort),
-				Remote:   fmt.Sprintf("%s:%d", virtualMachineIP, apiPort),
+				Local:    net.JoinHostPort(localIP, apiPort),
+				Remote:   net.JoinHostPort(virtualMachineIP, apiPort),
 			},
 			types.ExposeRequest{
 				Protocol: "tcp",
-				Local:    fmt.Sprintf(":%d", httpsPort),
-				Remote:   fmt.Sprintf("%s:%d", virtualMachineIP, httpsPort),
+				Local:    fmt.Sprintf(":%s", httpsPort),
+				Remote:   net.JoinHostPort(virtualMachineIP, httpsPort),
 			},
 			types.ExposeRequest{
 				Protocol: "tcp",
-				Local:    fmt.Sprintf(":%d", httpPort),
-				Remote:   fmt.Sprintf("%s:%d", virtualMachineIP, httpPort),
+				Local:    fmt.Sprintf(":%s", httpPort),
+				Remote:   net.JoinHostPort(virtualMachineIP, httpPort),
 			})
 	case crcPreset.Podman:
 		exposeRequest = append(exposeRequest,
 			types.ExposeRequest{
 				Protocol: "tcp",
-				Local:    fmt.Sprintf("%s:%d", localIP, cockpitPort),
-				Remote:   fmt.Sprintf("%s:%d", virtualMachineIP, cockpitPort),
+				Local:    net.JoinHostPort(localIP, cockpitPort),
+				Remote:   net.JoinHostPort(virtualMachineIP, cockpitPort),
 			},
 			types.ExposeRequest{
 				Protocol: "unix",
 				Local:    constants.GetHostDockerSocketPath(),
-				Remote:   fmt.Sprintf("ssh-tunnel://core@%s:%d/run/podman/podman.sock?key=%s", virtualMachineIP, internalSSHPort, constants.GetPrivateKeyPath()),
+				Remote:   fmt.Sprintf("ssh-tunnel://core@%s:%s/run/podman/podman.sock?key=%s", virtualMachineIP, internalSSHPort, constants.GetPrivateKeyPath()),
```

Comment:
This should be using net.JoinHostPort as well

Reply:
I believe you're right. I would also recommend constructing a https://pkg.go.dev/net/url#URL and setting its fields as necessary then stringifying at API boundaries. It'll handle all the weird edge cases.